File: l-umi/Week_21/PGS_산_모양_타일링.py
def solution(n, tops):
    
    a = [0]*n # 정삼각형으로 끝나는 경우
    b = [0]*n # 마름모로 끝나는 경우
    
    a[0] = 2 + tops[0]
    b[0] = 1
    
    # 점화식
    # b[i] = a[i-1]+b[i-1]
    # n == 1일때
    # a[i] = 3*a + 2*b
    # n == 0일때
    # a[i] = 2*a + b
    for i in range(1, n):
        a[i] = ( a[i-1] * (2+tops[i]) + b[i-1] * (1+tops[i]) ) % 10007
        b[i] = ( a[i-1]+b[i-1] ) % 10007
        
    return ( a[-1]+b[-1] ) % 10007

File: sickbirdd/Week_21/PGS_산_모양_타일링.java
class Solution {
	final static int MOD = 10007;
	static int[][] dp;
	public static int foo(int[] tops, int index, int type) {
		if(index == 0) return 1;
		if(dp[index][type] > 0) return dp[index][type];
		int cnt = tops[index - 1] == 1 ? 4 : 3;
		for(int i = 0; i < cnt; i++) {
			if(type == 1 && i != 2) {
				dp[index][type] += foo(tops, index - 1, i) % MOD;
			}
			else if(type != 1){
				dp[index][type] += foo(tops, index - 1, i) % MOD;
			}
		}
		return dp[index][type] % MOD;
	}
	public int solution(int n, int[] tops) {
		int ans = 0, cnt = tops[n - 1] == 1 ? 4 : 3;
		dp = new int[n][4];
		for(int i = 0; i < cnt; i++ ) {
			ans += foo(tops, n - 1, i);
		}
		return ans % MOD;
	}
}